## Task 8 
***
#### Explantion about task 1 to 7

All the task were realizated at the terminal of my computer or the codespace terminal and VSC as demonstrated in the class.

***
#### Task 1: Create Directory Structure

Was for created a folder in the repository the name 'data' and inside this folder created 2 files named 'timestamps' and 'weather'. 

I did this using the VSC. 

***
#### Task 2: Timestamps

Navigate to the data/timestamps directory. Use the date command to output the current date and time, appending the output to a file named now.txt. Make sure to use the >> operator to append (not overwrite) the file. Repeat this step ten times, then use the more command to verify that now.txt has the expected content.

I had enter in the terminal of the computer and followed the comands I had used to do the task.

##### Terminal.
:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>for /l %x in (1, 1, 10) do @echo %date% %time% >> now.txt

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>powershell -Command "Get-Date -Format 'yyyyMMdd_HHmmss'" >> formatted.txt

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>
C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>date /?
Displays or sets the date.

DATE [/T | date]

Type DATE without parameters to display the current date setting and
a prompt for a new one.  Press ENTER to keep the same date.

If Command Extensions are enabled the DATE command supports

***
#### Task 3: Formatting Timestamps

Run the date command again, but this time format the output using YYYYmmdd_HHMMSS (e.g., 20261114_130003 for 1:00:03 PM on November 14, 2026). Refer to the date man page (using man date) for more formatting options. (Press q to exit the man page). Append the formatted output to a file named formatted.txt.

As the task asked. I had done in the terminal using the comands below 

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>powershell -Command "Get-Date -Format 'Your_Format_String'"
Your_3or48aA_SArinA.D.

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>cd C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>powershell -Command "New-Item -ItemType File -Name ('{0:yyyyMMdd_HHmmss}.txt' -f (Get-Date))"


    Directory: C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastr
    ucture-1\data\timestamps

| Mode  | LastWriteTime     | Length | Name       |
|-------|-------------------|--------|------------|
| -a----| 17/12/2024 11:27  | 0      | weather.sh |


C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\timestamps>


C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data>mkdir weather

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data>cd weather

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\weather>curl -o weather.json https://prodapi.metweb.ie/observations/athenry/today

| % Total | % Received | % Xferd | Average Speed | Time | Time | Time | Current |
|---------|------------|---------|---------------|------|------|------|---------|
| Dload   | Upload     | Total   | Spent         | Left | Speed|      |         |
| 100     | 3348       | 0       | 3348          | 0    | 0    | 2391 | 0       |
| --:--:--| 0:00:01    | --:--:--| 2394          |      |      |      |         |

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\weather>


***
#### Task 4: Create Timestamped Files


Use the touch command to create an empty file with a name in the YYYYmmdd_HHMMSS.txt format. You can achieve this by embedding your date command in backticks ` into the touch command. You should no longer use redirection (>>) in this step.

I had use this comand 

New-Item -Path . -Name "$(Get-Date -Format 'yyyyMMdd_HHmmss').txt" -ItemType "File"        
>> 


    Directory: C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1 
    \data\timestamps

| Mode  | LastWriteTime     | Length | Name       |
|-------|-------------------|--------|------------|
| -a----| 17/12/2024 11:27  | 0      | weather.sh |

***

#### Task 5: Download Today's Weather Data


Change to the data/weather directory. Download the latest weather data for the Athenry weather station from Met Eireann using wget. Use the -O <filename> option to save the file as weather.json. The data can be found at this URL:
https://prodapi.metweb.ie/observations/athenry/today.


Using the terminal this comands below.
C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\weather>curl -o weather.json https://prodapi.metweb.ie/observations/athenry/today

| % Total | % Received | % Xferd | Average Speed | Time | Time | Time | Current |
|---------|------------|---------|---------------|------|------|------|---------|
| Dload   | Upload     | Total   | Spent         | Left | Speed|      |         |
| 100     | 3348       | 0       | 3348          | 0    | 0    | 2391 | 0       |
| --:--:--| 0:00:01    | --:--:--| 2394          |      |      |      |         |


***
#### Task 6: Timestamp the Data


Modify the command from Task 5 to save the downloaded file with a timestamped name in the format YYYYmmdd_HHMMSS.json.

C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1\data\weather>powershell -Command "$timestamp = (Get-Date -Format 'yyyyMMdd_HHmmss'); Invoke-WebRequest -Uri 'https://prodapi.metweb.ie/observations/athenry/today' -OutFile ($timestamp + '.json')"

***
#### Task 7: Write the Script


Write a bash script called weather.sh in the root of your repository. This script should automate the process from Task 6, saving the weather data to the data/weather directory. Make the script executable and test it by running it.

Directory: C:\Users\Administrator\Desktop\computer-infrastructure\computer-infrastructure-1 
    \data\timestamps


| Mode | LastWriteTime | Length | Name |
|------|---------------|--------|------|
| -a---- | 17/12/2024 11:27 | 0 | weather.sh |


***
## Task 9 Pands

In your weather.ipynb notebook, use the pandas function read_json() to load in any one of the weather data files you have downloaded with your script. Examine and summarize the data. Use the information provided data.gov.ie to write a short explanation of what the data set contains.

### Weather Analisys 

In [35]:
# Import the pandas and json libraries
import pandas as pd
import json


In [36]:
# Import os library
import os

# List files in the data/weather directory
weather_files = os.listdir('C:/Users/Administrator/Desktop/computer-infrastructure/computer-infrastructure-1/data/weather')
print(weather_files)


['20241114_110530.json', '20241114_111743.json', 'weather.json', 'weather.ps1', 'weather.sh']


In [37]:
# Read the first weather data file
weather_data = pd.read_json('C:/Users/Administrator/Desktop/computer-infrastructure/computer-infrastructure-1/data/weather/20241114_110530.json')


In [38]:
# Load the JSON file
with open('C:/Users/Administrator/Desktop/computer-infrastructure/computer-infrastructure-1/data/weather/20241114_110530.json') as f:
    data = json.load(f)

# Print the content of the JSON file
print(json.dumps(data, indent=4))


[
    {
        "name": "Athenry",
        "temperature": "9",
        "symbol": "15n",
        "weatherDescription": "Fog / Mist",
        "text": "\"Fog thickening\"",
        "windSpeed": "2",
        "windGust": "-",
        "cardinalWindDirection": "SW",
        "windDirection": 225,
        "humidity": " 100 ",
        "rainfall": " 0.1 ",
        "pressure": "1038",
        "dayName": "Thursday",
        "date": "14-11-2024",
        "reportTime": "00:00"
    },
    {
        "name": "Athenry",
        "temperature": "9",
        "symbol": "15n",
        "weatherDescription": "Fog / Mist",
        "text": "\"Fog thickening\"",
        "windSpeed": "2",
        "windGust": "-",
        "cardinalWindDirection": "W",
        "windDirection": 270,
        "humidity": " 100 ",
        "rainfall": " 0.0 ",
        "pressure": "1038",
        "dayName": "Thursday",
        "date": "14-11-2024",
        "reportTime": "01:00"
    },
    {
        "name": "Athenry",
        "temperature":

In [39]:
# Display the first few rows of the data
print(weather_data.head())

# Display summary statistics of the data
print(weather_data.describe())

# Display information about the data types and missing values
print(weather_data.info())


      name  temperature symbol weatherDescription              text windSpeed  \
0  Athenry            9    15n         Fog / Mist  "Fog thickening"         2   
1  Athenry            9    15n         Fog / Mist  "Fog thickening"         2   
2  Athenry           10    15n         Fog / Mist    "Fog thinning"         -   
3  Athenry           10    09n               Rain  "Moderate rain "         4   
4  Athenry           10    09n               Rain  "Moderate rain "         4   

  windGust cardinalWindDirection  windDirection  humidity  rainfall  pressure  \
0        -                    SW            225       100      0.10      1038   
1        -                     W            270       100      0.00      1038   
2        -                                    0        99      0.00      1037   
3        -                     N              0       100      0.01      1037   
4        -                     N              0        99      0.01      1037   

    dayName       date rep

In [40]:
# Display the column names
print("Column names:", weather_data.columns)

# Display the data types of each column
print("Data types:", weather_data.dtypes)

# Display the number of missing values in each column
print("Missing values:", weather_data.isnull().sum())


Column names: Index(['name', 'temperature', 'symbol', 'weatherDescription', 'text',
       'windSpeed', 'windGust', 'cardinalWindDirection', 'windDirection',
       'humidity', 'rainfall', 'pressure', 'dayName', 'date', 'reportTime'],
      dtype='object')
Data types: name                             object
temperature                       int64
symbol                           object
weatherDescription               object
text                             object
windSpeed                        object
windGust                         object
cardinalWindDirection            object
windDirection                     int64
humidity                          int64
rainfall                        float64
pressure                          int64
dayName                          object
date                     datetime64[ns]
reportTime                       object
dtype: object
Missing values: name                     0
temperature              0
symbol                   0
weatherDescription  

***
##### Summarize the key information:

- Location: The name column contains the location name.

- Temperature: The temperature column contains the temperature in Celsius.

- Weather Description: The weatherDescription column contains a description of the weather.

- Wind Speed: The windSpeed column contains the wind speed.

- Humidity: The humidity column contains the humidity percentage.

- Rainfall: The rainfall column contains the amount of rainfall.

- Pressure: The pressure column contains the atmospheric pressure.

- Date: The date column contains the date of the weather report.

- Report Time: The reportTime column contains the time of the weather report.
***

In [41]:
# Sumarize the weather data

summary = {
    "Location": weather_data['name'][0],
    "Temperature (C)": weather_data['temperature'][0],
    "Weather Description": weather_data['weatherDescription'][0],
    "Wind Speed": weather_data['windSpeed'][0],
    "Humidity (%)": weather_data['humidity'][0],
    "Rainfall (mm)": weather_data['rainfall'][0],
    "Pressure (hPa)": weather_data['pressure'][0],
    "Date": weather_data['date'][0],
    "Report Time": weather_data['reportTime'][0]
}

print(summary)


{'Location': 'Athenry', 'Temperature (C)': np.int64(9), 'Weather Description': 'Fog / Mist', 'Wind Speed': '2', 'Humidity (%)': np.int64(100), 'Rainfall (mm)': np.float64(0.1), 'Pressure (hPa)': np.int64(1038), 'Date': Timestamp('2024-11-14 00:00:00'), 'Report Time': '00:00'}


***
##### Weather Data Set Explanation

The weather data set provides detailed information about the current weather conditions for a specific location. The data includes the following key attributes:

- Location: The name of the location where the weather data was recorded (e.g., Athenry).

- Temperature (C): The temperature in Celsius (e.g., 9°C).

- Weather Description: A description of the weather conditions (e.g., Fog / Mist).

- Wind Speed: The speed of the wind (e.g., 2 kph).

- Humidity (%): The percentage of humidity in the air (e.g., 100%).

- Rainfall (mm): The amount of rainfall in millimeters (e.g., 0.1 mm).

- Pressure (hPa): The atmospheric pressure in hectopascals (e.g., 1038 hPa).

- Date: The date when the weather data was recorded (e.g., 2024-11-14).

- Report Time: The time when the weather data was recorded (e.g., 00:00).

***
# End